### PR TITLE
PS-5773 : crash in create temporary with encryption on

### PR DESCRIPTION
**Pull request by satya-bodapati**
Problem:
--------
When temporary tablespace is encrypted dynamically
(SET GLOBAL innodb_temp_tablespace_encrypt=ON), CREATE TEMPORARY TABLE
or INSERTs into that the table crashes the server

Assertion failure because tablespace flags mismatch.
In-memory flags indicate presence of encryption bit where on disk Page0
flags do not.

Fix:
----
When encryption is turned on, update the Page 0 flags with encryption bit.

### CODE REVIEW

File: mysql-test/suite/innodb/t/temp_table_encrypt.test [hunk 1]
```diff
@@ -94,4 +94,23 @@ INSERT INTO t01 VALUES (1), (2), (3);
 SET GLOBAL innodb_temp_tablespace_encrypt = OFF;
```
Comment (laurynas-biveinis) [inline mysql-test/suite/innodb/t/temp_table_encrypt.test]: Now redundant with the restart

File: mysql-test/suite/innodb/t/temp_table_encrypt.test [hunk 2]
```diff
@@ -94,4 +94,23 @@ INSERT INTO t01 VALUES (1), (2), (3);
...
 
 --source include/wait_until_count_sessions.inc
```
Comment (laurynas-biveinis) [inline mysql-test/suite/innodb/t/temp_table_encrypt.test]: Now redundant with the restart

File: storage/innobase/fil/fil0fil.cc [hunk 3]
```diff
@@ -7932,17 +7932,19 @@ fil_get_compression(
...
+@param[in] acquire_mutex	if true acquire fil_sys mutex, else false
 @return DB_SUCCESS or error code */
 dberr_t
 fil_set_encryption(
 	ulint			space_id,
 	Encryption::Type	algorithm,
 	byte*			key,
 	byte*			iv,
-	bool aquire_mutex)
+	bool			aquire_mutex)
```
Comment (laurynas-biveinis) [inline storage/innobase/fil/fil0fil.cc]: s/aquire_mutex/acquire_mutex now that you touch it?

File: storage/innobase/fil/fil0fil.cc [hunk 4]
```diff
@@ -7932,17 +7932,19 @@ fil_get_compression(
...
 {
-	if (aquire_mutex)
+	if (aquire_mutex) {
 		mutex_enter(&fil_system->mutex);
+	}
 
 	fil_space_t*	space = fil_space_get_by_id(space_id);
```
Comment (laurynas-biveinis) [inline storage/innobase/fil/fil0fil.cc:7950]: if (space == NULL) {
  if (acquire_mutex) mutex_exit(&fil_system->mutex);
  return(DB_NOT_FOUND);
}

File: storage/innobase/include/fil0fil.h [hunk 5]
```diff
@@ -1766,25 +1766,15 @@ fil_get_compression(
...
-	MY_ATTRIBUTE((warn_unused_result));
-*/
 dberr_t
 fil_set_encryption(
 	ulint			space_id,
 	Encryption::Type	algorithm,
 	byte*			key,
 	byte*			iv,
-        bool aquire_mutex = true)
+        bool			aquire_mutex = true)
```
Comment (laurynas-biveinis) [inline storage/innobase/include/fil0fil.h]: Formatting
s/aquire/acquire

File: storage/innobase/fsp/fsp0fsp.cc [hunk 6]
```diff
@@ -1139,7 +1139,9 @@ fsp_enable_encryption(
```
Comment (laurynas-biveinis) [inline storage/innobase/fsp/fsp0fsp.cc:1139]: at the top ut_ad(mutex_own(&fil_system->mutex)); ?

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: mysql-test/suite/innodb/t/temp_table_encrypt.test (hunk 1 above)
Comment (satya-bodapati) [reply mysql-test/suite/innodb/t/temp_table_encrypt.test]: but this is after restart, at the end of test right?

File: storage/innobase/fsp/fsp0fsp.cc (hunk 6 above)
Comment (satya-bodapati) [reply storage/innobase/fsp/fsp0fsp.cc:1139]: actually fil_sys is not acquired by the caller. There seems to be a bug below with extra "mutex_exit()" on failure

File: mysql-test/suite/innodb/t/temp_table_encrypt.test (hunk 1 above)
Comment (satya-bodapati) [reply mysql-test/suite/innodb/t/temp_table_encrypt.test]: never mind, this is at other place

File: storage/innobase/fsp/fsp0fsp.cc (hunk 6 above)
Comment (satya-bodapati) [reply storage/innobase/fsp/fsp0fsp.cc:1139]: removed that extra mutex_exit()

**Comment by satya-bodapati** [comment]
Ready for review

**Review by laurynas-biveinis — APPROVED** [review]

**Pull request merged**